[AY1920S1-CS2113T-W17-3] OwlMoney

shamussy opened the pull request:
@shamussy
@briantkm
@terencetanwt
@mong-nus
@valerietyj

jinyao-lee commented on src/main/java/owlmoney/Main.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Initializes a new OwlMoney session.
+     */
+    private Main() {
```
Method name should start with a verb (e.g.: `private initializeSession(...)`)

shamussy replied:
Fixed in PR 108, I am not using hashtags to prevent tagging other groups.

jinyao-lee commented on src/main/java/owlmoney/Main.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * The main class.
+ */
+
```
Should not have a blank line here

shamussy replied:
Fixed in PR 108, I am not using hashtags to prevent tagging other groups.

jinyao-lee commented on src/main/java/owlmoney/Main.java:
```diff
@@ -0,0 +1,73 @@
+        } catch (FileNotFoundException e) {
+            ui.printError("Could not read tasks from disk, will start with empty file");
+            tasks = new TaskList();
+        }*/
```
Commented code does not help the readability and should be deleted

shamussy replied:
Fixed in PR 130, these codes were temporary placeholders before  I implemented persistent storage for OwlMoney.

jinyao-lee commented on src/main/java/owlmoney/Main.java:
```diff
@@ -0,0 +1,73 @@
+                if (hasExited) {
+                    break;
+                }
+
```
Unnecessary blank line

shamussy replied:
Fixed in PR 108, I am not using hashtags to prevent tagging other groups.

jinyao-lee commented on src/main/java/owlmoney/logic/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+     * @param ui      Ui of OwlMoney.
+     * @return True if OwlMoney should terminate after execution.
+     */
+    public abstract boolean execute(Profile profile, Ui ui) throws BankException,
```
Docs comments for methods should start with a verb

shamussy replied:
Fixed in PR 110 and 111, I am not using hashtags to prevent tagging other groups.

jinyao-lee commented on src/main/java/owlmoney/logic/command/PlaceHolderEmptyCommand.java:
```diff
@@ -0,0 +1,17 @@
+import owlmoney.model.profile.Profile;
+import owlmoney.ui.Ui;
+
+public class PlaceHolderEmptyCommand extends Command {
```
"What's the purpose of this class?" --> Each class should have a Javadocs comment to explain the purpose of this class

shamussy replied:
Fixed in PR 131, was a placeholder to prevent errors from always appearing since we have not implemented the command previously

jinyao-lee commented on src/main/java/owlmoney/logic/command/bank/DeleteInvestmentCommand.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public class DeleteInvestmentCommand extends Command {
+    private final String bankName;
+    private static final String INVESTMENT = "investment";
```
`INVESTMENT = "investment"` doesn't explain the purpose of this variable. Is this used as a bank account type? If so, rename it such that readers will know that you are trying to convey that this is an account type

shamussy replied:
Fixed in PR 210

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseCommand.java:
```diff
@@ -0,0 +1,88 @@
+public class ParseCommand extends Parser {
+    private ParseType parseType = new ParseType();
+    private final Scanner scanner = new Scanner(System.in);
+    private static final String[] COMMAND_KEYWORDS = new String[] {
```
The contents of your `COMMAND_KEYWORDS` are equivalent to **magic numbers**.

jinyao-lee replied:
```
Hi @jylee-git ,

With regards to your:

The contents of your COMMAND_KEYWORDS are equivalent to magic numbers.

I thought we did the right thing by putting all the supposedly magic strings in a declared constant with private static final?

Regards
Shamus Neo
```

Maybe let me rephrase it: It's equivalent to dumping a collection of magic numbers inside.

Because just dumping these `"/help", "/undo", "/add", "/edit", "/delete", "/list", "/find", "/transfer", "/exit"` don't explain what are these values about.

So for this sort of case, you should declare static final variables for each these commands, then dump the created static final variables in to your `private static final List<String> COMMAND_KEYWORD_LISTS`.

Something like this:

```java
private static final String CMD_HELP = "/help";
//...
private static final List<String> CMD_KEYWRDS = Arrays.asList(CMD_HELP, CMD_ADD, CMD_otherCmdsYouWantToPut);
```

shamussy replied:
Alright noted, will correct them

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseCommand.java:
```diff
@@ -0,0 +1,88 @@
+        case "/list":
+            return parseType.parseData(command, data);
+        case "/exit":
+            return new ExitCommand();
```
Magic numbers

shamussy replied:
Fixed in PR 210

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseRawData.java:
```diff
@@ -0,0 +1,85 @@
+ * Represents the third layer of parsing for raw data after removing command and type.
+ * This forms the baseline raw data parsing that specific data level parsers can extend from.
+ */
+
```
Unnecessary blank line. The javadoc comments should be directly above your class or method without any blank lines

shamussy replied:
Fixed in PR 108, 110 and 111, I am not using hashtags to prevent tagging other groups.

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseRawData.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    private void checkEachKeyword(String keyword, String[] splitArray) throws ParserException {
```
What is checkEachKeyword?

shamussy replied:
Resolved in PR 225

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseRawData.java:
```diff
@@ -0,0 +1,85 @@
+        return parameter;
+    }
+
+    private void checkDuplicateKeywords(String[] splitArray, String[] keywordList) throws ParserException {
```
The method name trips the reader (tripped me) as it is vague. Use a more relevant name like `assertNoDuplicateKeywords`. 

A note in general (may/may not apply to here): you should provide javadoc comments for private methods as well if they are non-trivial.

shamussy replied:
Resolved in PR 225

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseRawData.java:
```diff
@@ -0,0 +1,85 @@
+                return i;
+            }
+        }
+        return -1;
```
Magic numbers

shamussy replied:
Resolved in PR 225

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseType.java:
```diff
@@ -0,0 +1,291 @@
+     * @param command The extracted first field from the initial user input that determines the command.
+     * @throws ParserException if the user wants to delete his profile.
+     */
+    private void isDeleteProfile(String command) throws ParserException {
```
`is` keyword should only be used if the method returns a boolean.

Use keywords such as assert or validate.

shamussy replied:
Resolved in PR 225

jinyao-lee commented on src/main/java/owlmoney/logic/parser/ParseType.java:
```diff
@@ -0,0 +1,291 @@
+        default:
+            throw new ParserException("You entered an invalid type");
+        }
+    }
```
This method is wayyy too lengthy... Magic numbers as well. Find a way to shorten this!

![image](https://user-images.githubusercontent.com/35756209/67156931-7577ca80-f357-11e9-9818-3fb9cb74fcad.png)

shamussy replied:
Fixed in PR 222

jinyao-lee commented on src/main/java/owlmoney/logic/regex/RegexUtil.java:
```diff
@@ -0,0 +1,148 @@
+     * @return the result of the check on whether it fulfills the criteria.
+     */
+    public static boolean regexCheckInterestRate(String input) {
+        final String interestRegex = "^\\s*(?=.*[1-9])\\d{1,2}(\\.\\d{1,2})?$";
```
Line 56 (seems) to have the same regex as well. Make it a class level variable instead.

I.e.: avoid duplicate codes

shamussy replied:
Fixed in PR 210

jinyao-lee commented on src/main/java/owlmoney/model/bank/BankList.java:
```diff
@@ -0,0 +1,589 @@
+        for (int i = 0; i < bankLists.size(); i++) {
+            if (bankLists.get(i).getAccountName().equals(bankName)
+                    && "investment".equals(bankLists.get(i).getType())) {
+                if (!(newName.isEmpty() || newName.isBlank())) {
```
`asd.isBlank()` is quivalent to `asd.trim().isempty()`.

You can eliminate the use of isEmpty in this case. same as for those above

shamussy replied:
Fixed in PR 218

jinyao-lee commented on src/main/java/owlmoney/model/transaction/TransactionList.java:
```diff
@@ -0,0 +1,315 @@
+     * @return New amount of the expenditure.
+     * @throws TransactionException If incorrect date format.
+     */
+    public double editEx(int expNum, String desc, String amount, String date, String category, Ui ui)
```
Avoid such shorthands for your method name. This trips the readers

shamussy replied:
Fixed before PR 210

jinyao-lee reviewed (commented):
Overall good job (with the help of Travis weeding out other checkstyle issues).

Note that the comments are not exhaustive (whatever I commented also applies to those below/above).

Things to note (the more serious issues):
- Avoid long method (one of your method is > 100 lines long!). Readers will have a hard time reading the code and you will have a tough time debugging that method.
- Avoid magic numbers. This problem occurred in many parts of your code.

shamussy commented:
Hi Jin Yao,

With regards to your:

The contents of your COMMAND_KEYWORDS are equivalent to magic numbers.

I thought we did the right thing by putting all the supposedly magic strings in a declared constant with private static final?

Regards
Shamus Neo
________________________________
From: Lee Jin Yao <notifications@github.com>
Sent: Sunday, 20 October 2019 5:01 PM
To: nusCS2113-AY1920S1/PersonalAssistant-Duke <PersonalAssistant-Duke@noreply.github.com>
Cc: Shamus Neo Zhi Kai <e0310778@u.nus.edu>; Mention <mention@noreply.github.com>
Subject: Re: [nusCS2113-AY1920S1/PersonalAssistant-Duke] [AY1920S1-CS2113T-W17-3] OwlMoney (#26)


@jylee-git commented on this pull request.

Overall good job (with the help of Travis weeding out other checkstyle issues).

Note that the comments are not exhaustive (whatever I commented also applies to those below/above).

Things to note (the more serious issues):

  *   Avoid long method (one of your method is > 100 lines long!). Readers will have a hard time reading the code and you will have a tough time debugging that method.
  *   Avoid magic numbers. This problem occurred in many parts of your code.

________________________________

In src/main/java/owlmoney/Main.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336766393>:

> +
+/**
+ * The main class.
+ */
+
+class Main {
+
+    private Ui ui;
+    private ParseCommand parser;
+    private Profile profile;
+    //private Storage storage;
+
+    /**
+     * Initializes a new OwlMoney session.
+     */
+    private Main() {


Method name should start with a verb (e.g.: private initializeSession(...))

________________________________

In src/main/java/owlmoney/Main.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336766426>:

> +
+import owlmoney.logic.command.Command;
+import owlmoney.logic.parser.ParseCommand;
+import owlmoney.model.card.exception.CardException;
+import owlmoney.logic.parser.exception.ParserException;
+import owlmoney.model.bank.exception.BankException;
+import owlmoney.model.bond.exception.BondException;
+import owlmoney.model.goals.exception.GoalsException;
+import owlmoney.model.profile.Profile;
+import owlmoney.model.transaction.exception.TransactionException;
+import owlmoney.ui.Ui;
+
+/**
+ * The main class.
+ */
+


Should not have a blank line here

________________________________

In src/main/java/owlmoney/Main.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336766546>:

> +        /*storage = new Storage("data/data.txt");
+        try {
+            tasks = new TaskList(storage.readFile());
+        } catch (FileNotFoundException e) {
+            ui.printError("Could not read tasks from disk, will start with empty file");
+            tasks = new TaskList();
+        }*/


Commented code does not help the readability and should be deleted

________________________________

In src/main/java/owlmoney/Main.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336766570>:

> +        boolean hasExited = false;
+        //Temporary do this chunk
+        ui.firstTimeRun();
+        Scanner scanner = new Scanner(System.in);
+        String username = scanner.nextLine();
+        profile = new Profile(username);
+        ui.greet(profile.profileGetUsername());
+        // until above this line
+        while (parser.hasNextLine()) {
+            try {
+                Command command = parser.parseLine();
+                hasExited = command.execute(profile, ui);
+                if (hasExited) {
+                    break;
+                }
+


Unnecessary blank line

________________________________

In src/main/java/owlmoney/logic/command/Command.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336766725>:

> +    /**
+     * Abstract method where each command type implements execution code.
+     *
+     * @param profile Profile of the user.
+     * @param ui      Ui of OwlMoney.
+     * @return True if OwlMoney should terminate after execution.
+     */
+    public abstract boolean execute(Profile profile, Ui ui) throws BankException,


Docs comments for methods should start with a verb

________________________________

In src/main/java/owlmoney/logic/command/PlaceHolderEmptyCommand.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336766842>:

> @@ -0,0 +1,17 @@
+package owlmoney.logic.command;
+
+import owlmoney.model.profile.Profile;
+import owlmoney.ui.Ui;
+
+public class PlaceHolderEmptyCommand extends Command {


"What's the purpose of this class?" --> Each class should have a Javadocs comment to explain the purpose of this class

________________________________

In src/main/java/owlmoney/logic/command/bank/DeleteInvestmentCommand.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336767034>:

> @@ -0,0 +1,37 @@
+package owlmoney.logic.command.bank;
+
+import owlmoney.logic.command.Command;
+import owlmoney.model.bank.exception.BankException;
+import owlmoney.model.profile.Profile;
+import owlmoney.ui.Ui;
+
+/**
+ * DeleteInvestmentCommand contains the execution function to delete an investment object.
+ */
+public class DeleteInvestmentCommand extends Command {
+    private final String bankName;
+    private static final String INVESTMENT = "investment";


INVESTMENT = "investment" doesn't explain the purpose of this variable. Is this used as a bank account type? If so, rename it such that readers will know that you are trying to convey that this is an account type

________________________________

In src/main/java/owlmoney/logic/parser/ParseCommand.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336767824>:

> +import java.util.List;
+import java.util.Scanner;
+
+import owlmoney.logic.command.Command;
+import owlmoney.logic.command.ExitCommand;
+import owlmoney.logic.parser.exception.ParserException;
+
+/**
+ * Represents the first instance of parsing user input.
+ * This determines what type of command the user has entered.
+ */
+
+public class ParseCommand extends Parser {
+    private ParseType parseType = new ParseType();
+    private final Scanner scanner = new Scanner(System.in);
+    private static final String[] COMMAND_KEYWORDS = new String[] {


The contents of your COMMAND_KEYWORDS are equivalent to magic numbers.

________________________________

In src/main/java/owlmoney/logic/parser/ParseCommand.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336768040>:

> +        case "/add":
+            // Fallthrough
+        case "/delete":
+            // Fallthrough
+        case "/edit":
+            // Fallthrough
+        case "/list":
+            return parseType.parseData(command, data);
+        case "/exit":
+            return new ExitCommand();


Magic numbers

________________________________

In src/main/java/owlmoney/logic/parser/ParseRawData.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336768076>:

> @@ -0,0 +1,85 @@
+package owlmoney.logic.parser;
+
+import owlmoney.logic.parser.exception.ParserException;
+
+/**
+ * Represents the third layer of parsing for raw data after removing command and type.
+ * This forms the baseline raw data parsing that specific data level parsers can extend from.
+ */
+


Unnecessary blank line. The javadoc comments should be directly above your class or method without any blank lines

________________________________

In src/main/java/owlmoney/logic/parser/ParseRawData.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336768202>:

> +
+        int startIndex = findStartIndex(splitTypeSpecificArguments, keyword);
+        if (startIndex >= 0) {
+            int endIndex = findEndIndex(startIndex, splitTypeSpecificArguments, keywordList);
+            parameter = concatenateParameter(startIndex, endIndex, splitTypeSpecificArguments);
+        }
+        return parameter;
+    }
+
+    private void checkDuplicateKeywords(String[] splitArray, String[] keywordList) throws ParserException {
+        for (int i = 0; i < keywordList.length; i++) {
+            checkEachKeyword(keywordList[i], splitArray);
+        }
+    }
+
+    private void checkEachKeyword(String keyword, String[] splitArray) throws ParserException {


What is checkEachKeyword?

________________________________

In src/main/java/owlmoney/logic/parser/ParseRawData.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336768403>:

> +     * @throws ParserException if duplicated keywords are detected.
+     */
+    public String extractParameter(String data, String keyword, String[] keywordList) throws ParserException {
+        String[] splitTypeSpecificArguments = data.split(" ");
+        String parameter = "";
+        checkDuplicateKeywords(splitTypeSpecificArguments, keywordList);
+
+        int startIndex = findStartIndex(splitTypeSpecificArguments, keyword);
+        if (startIndex >= 0) {
+            int endIndex = findEndIndex(startIndex, splitTypeSpecificArguments, keywordList);
+            parameter = concatenateParameter(startIndex, endIndex, splitTypeSpecificArguments);
+        }
+        return parameter;
+    }
+
+    private void checkDuplicateKeywords(String[] splitArray, String[] keywordList) throws ParserException {


The method name trips the reader (tripped me) as it is vague. Use a more relevant name like assertNoDuplicateKeywords.

A note in general (may/may not apply to here): you should provide javadoc comments for private methods as well if they are non-trivial.

________________________________

In src/main/java/owlmoney/logic/parser/ParseRawData.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336768427>:

> +            if (splitArray[i].equals(keyword)) {
+                keywordCounter++;
+            }
+            if (keywordCounter > 1) {
+                throw new ParserException("Duplicate " + keyword + " found!");
+            }
+        }
+    }
+
+    private int findStartIndex(String[] splitTypeSpecificArguments, String keyword) {
+        for (int i = 0; i < splitTypeSpecificArguments.length; i++) {
+            if (splitTypeSpecificArguments[i].equals(keyword)) {
+                return i;
+            }
+        }
+        return -1;


Magic numbers

________________________________

In src/main/java/owlmoney/logic/parser/ParseType.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336768581>:

> +        String rawData;
+        if ("/list".equals(command)) {
+            rawData = removeListFirstField(data, type);
+        } else {
+            rawData = removeFirstField(data, type);
+        }
+        return parseTypeMenu(command, type, rawData);
+    }
+
+    /**
+     * Checks if the user wants to delete profile.
+     *
+     * @param command The extracted first field from the initial user input that determines the command.
+     * @throws ParserException if the user wants to delete his profile.
+     */
+    private void isDeleteProfile(String command) throws ParserException {


is keyword should only be used if the method returns a boolean.

Use keywords such as assert or validate.

________________________________

In src/main/java/owlmoney/logic/parser/ParseType.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336768702>:

> +                deleteExp.fillHashTable();
+                deleteExp.checkParameter();
+                return deleteExp.getCommand();
+            } else if ("/edit".equals(command)) {
+                ParseExpenditure editExp = new ParseEditExpenditure(rawData, BANK);
+                editExp.fillHashTable();
+                editExp.checkParameter();
+                return editExp.getCommand();
+            }
+            throw new ParserException("You entered an invalid type for bank expenditure");
+        case "/cardexpenditure":
+            if ("/add".equals(command)) {
+                ParseExpenditure addExp = new ParseAddExpenditure(rawData, CARD);
+                addExp.fillHashTable();
+                addExp.checkParameter();
+                return addExp.getCommand();
+            } else if ("/list".equals(command)) {
+                ParseExpenditure listExp = new ParseListExpenditure(rawData, CARD);
+                listExp.fillHashTable();
+                listExp.checkParameter();
+                return listExp.getCommand();
+            } else if ("/delete".equals(command)) {
+                ParseExpenditure deleteExp = new ParseDeleteExpenditure(rawData, CARD);
+                deleteExp.fillHashTable();
+                deleteExp.checkParameter();
+                return deleteExp.getCommand();
+            } else if ("/edit".equals(command)) {
+                ParseExpenditure editExp = new ParseEditExpenditure(rawData, CARD);
+                editExp.fillHashTable();
+                editExp.checkParameter();
+                return editExp.getCommand();
+            }
+            throw new ParserException("You entered an invalid type for card expenditure");
+        case "/deposit":
+            if ("/add".equals(command)) {
+                ParseDeposit addDep = new ParseAddDeposit(rawData);
+                addDep.fillHashTable();
+                addDep.checkParameter();
+                return addDep.getCommand();
+            } else if ("/list".equals(command)) {
+                ParseDeposit listDep = new ParseListDeposit(rawData);
+                listDep.fillHashTable();
+                listDep.checkParameter();
+                return listDep.getCommand();
+            } else if ("/delete".equals(command)) {
+                ParseDeposit deleteDep = new ParseDeleteDeposit(rawData);
+                deleteDep.fillHashTable();
+                deleteDep.checkParameter();
+                return deleteDep.getCommand();
+            } else if ("/edit".equals(command)) {
+                ParseDeposit editDep = new ParseEditDeposit(rawData);
+                editDep.fillHashTable();
+                editDep.checkParameter();
+                return editDep.getCommand();
+            }
+            throw new ParserException("You entered an invalid type for deposit");
+        case "/card":
+            if ("/add".equals(command)) {
+                ParseCard addCard = new ParseAddCard(rawData);
+                addCard.fillHashTable();
+                addCard.checkParameter();
+                return addCard.getCommand();
+            } else if ("/delete".equals(command)) {
+                ParseCard deleteCard = new ParseDeleteCard(rawData);
+                deleteCard.fillHashTable();
+                deleteCard.checkParameter();
+                return deleteCard.getCommand();
+            } else if ("/list".equals(command)) {
+                return new ListCardCommand();
+            } else if ("/edit".equals(command)) {
+                ParseCard editCard = new ParseEditCard(rawData);
+                editCard.fillHashTable();
+                editCard.checkParameter();
+                return editCard.getCommand();
+            }
+            throw new ParserException("You entered an invalid type for card");
+        case "/goals":
+            if ("/add".equals(command)) {
+                ParseGoals addGoals = new ParseAddGoals(rawData);
+                addGoals.fillHashTable();
+                addGoals.checkParameter();
+                return addGoals.getCommand();
+            } else if ("/delete".equals(command)) {
+                ParseGoals deleteGoals = new ParseDeleteGoals(rawData);
+                deleteGoals.fillHashTable();
+                deleteGoals.checkParameter();
+                return deleteGoals.getCommand();
+            } else if ("/edit".equals(command)) {
+                ParseGoals editGoals = new ParseEditGoals(rawData);
+                editGoals.fillHashTable();
+                editGoals.checkParameter();
+                return editGoals.getCommand();
+            } else if ("/list".equals(command)) {
+                return new ListGoalsCommand();
+            }
+            throw new ParserException("You entered an invalid type for goals");
+        default:
+            throw new ParserException("You entered an invalid type");
+        }
+    }


This method is wayyy too lengthy... Magic numbers as well. Find a way to shorten this!

[image]<https://user-images.githubusercontent.com/35756209/67156931-7577ca80-f357-11e9-9818-3fb9cb74fcad.png>

________________________________

In src/main/java/owlmoney/logic/regex/RegexUtil.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336769102>:

> +     */
+    public static boolean regexCheckBankAmount(String input) {
+        final String moneyRegex = "^\\s*(?=.*[0-9])\\d{1,9}(\\.\\d{1,2})?$";
+        return input.matches(moneyRegex);
+    }
+
+    /**
+     * Checks whether the interest rate entered is within the 100% limit.
+     * The first line of checks determines whether it is a 2 digit number up to 2 decimal places.
+     * The second line of checks determines whether it is less than 100%.
+     *
+     * @param input The user input that is subject to Regex checking.
+     * @return the result of the check on whether it fulfills the criteria.
+     */
+    public static boolean regexCheckInterestRate(String input) {
+        final String interestRegex = "^\\s*(?=.*[1-9])\\d{1,2}(\\.\\d{1,2})?$";


Line 56 seems to have the same regex as well. Make it a class level variable instead.

________________________________

In src/main/java/owlmoney/model/bank/BankList.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336769291>:

> +
+    /**
+     * Edits the investment account details.
+     *
+     * @param bankName Bank account to be edited.
+     * @param newName  New name of bank account.
+     * @param amount   New amount of bank account.
+     * @param ui       required for printing.
+     * @throws BankException If duplicate bank name found.
+     */
+    public void bankListEditInvestment(String bankName, String newName, String amount, Ui ui)
+            throws BankException {
+        for (int i = 0; i < bankLists.size(); i++) {
+            if (bankLists.get(i).getAccountName().equals(bankName)
+                    && "investment".equals(bankLists.get(i).getType())) {
+                if (!(newName.isEmpty() || newName.isBlank())) {


asd.isBlank() is quivalent to asd.trim().isempty().

You can eliminate the use of isEmpty in this case. same as for those above

________________________________

In src/main/java/owlmoney/model/transaction/TransactionList.java<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26#discussion_r336769437>:

> +        }
+    }
+
+    /**
+     * Edits the specific expenditure in the list.
+     *
+     * @param expNum   Transaction number of the expenditure.
+     * @param desc     New description of the expenditure.
+     * @param amount   New amount of the expenditure.
+     * @param date     New date of the expenditure.
+     * @param category New category of the expenditure.
+     * @param ui       required for printing.
+     * @return New amount of the expenditure.
+     * @throws TransactionException If incorrect date format.
+     */
+    public double editEx(int expNum, String desc, String amount, String date, String category, Ui ui)


Avoid such shorthands for your method name. This trips the readers

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/nusCS2113-AY1920S1/PersonalAssistant-Duke/pull/26?email_source=notifications&email_token=AIPTXEGA6P3CR4DPWGF25F3QPQM7BA5CNFSM4IVH3ZU2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIRGV5A#pullrequestreview-304245492>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AIPTXEDGMZ4GFV5C43P7MXTQPQM7BANCNFSM4IVH3ZUQ>.

shamussy commented:
Hi @jylee-git ,

With regards to your:

The contents of your COMMAND_KEYWORDS are equivalent to magic numbers.

I thought we did the right thing by putting all the supposedly magic strings in a declared constant with private static final?

Regards
Shamus Neo

jinyao-lee commented:
P.S.: If anything, comment directly on the ones that I've commented

![image](https://user-images.githubusercontent.com/35756209/67157391-e5894f00-f35d-11e9-9297-2254e993eb13.png)

TerenceTanWT committed:
Update incorrect heading in DG

shamussy committed:
Update PPP, DG UG, for typos and updated information to prepare for 1.4

TerenceTanWT committed:
Edit Card DG formatting

shamussy committed:
Merge pull request #203 from valerietyj/goalmidv1.4

Implement storage for achievements

shamussy committed:
Merge pull request #204 from BrianTKM/ManualTest

Manual test

TerenceTanWT committed:
Merge with master to check for conflicts

TerenceTanWT committed:
Merge pull request #205 from shamussy/documentation

Update UG, DG, PPP

TerenceTanWT committed:
Removed trailing white space

TerenceTanWT committed:
Removed trailing white space

TerenceTanWT committed:
Removed trailing white space

TerenceTanWT committed:
Added PPP

TerenceTanWT committed:
Added back part of DG that I removed

TerenceTanWT committed:
Added back part of DG that I removed

valerietyj committed:
Implemented untying of goals when linked to bank acc

valerietyj committed:
Update user guide and fix checkstyle

valerietyj committed:
Fix codacy error

valerietyj committed:
Fix codacy error

shamussy committed:
Merge pull request #206 from TerenceTanWT/DGManualTest

Added Instructions for Manual Testing for Card

shamussy committed:
Merge pull request #208 from valerietyj/goalmidv1.4

Allow user to untie savings account linked to goals

mong-nus committed:
add picture to UG for find and transfer

shamussy committed:
Implement primary logging functions and tested on profile

shamussy committed:
Fix EOF and magic strings

shamussy committed:
Major variable renaming to follow java conventions as well as minor bug fixes

shamussy committed:
Rearrange Manual Test in DG

mong-nus committed:
changes to picture for ug find and transfer

mong-nus committed:
remove picture

mong-nus committed:
add picture

mong-nus committed:
make amendment to image

mong-nus committed:
change image name

mong-nus committed:
fix naming issue of image in UG

mong-nus committed:
change image for PPP

mong-nus committed:
more picture for UG

mong-nus committed:
fix image issue

mong-nus committed:
remove wrong delete card image

mong-nus committed:
add correct delete card image

mong-nus committed:
remove photo from UG

mong-nus committed:
added better photo into UG

BrianTKM committed:
Merge pull request #210 from shamussy/loganddoc

Implement primary logging functions and tested on profile

mong-nus committed:
Merge branch 'master' into documentationPPP

Merge with upstream

shamussy committed:
Modify readme

shamussy committed:
Fix linking of ui photo

mong-nus committed:
amend manual testing error

TerenceTanWT committed:
Fixes bug #194 and #195 to prevent duplicated card bill and card expenditures overflow
Resolves #194
Resolves #195

Fixes bug #194 to prevent user from paying card bill for a particular month if it has already been paid before.
Fixes bug #195 to prevent user from adding card expenditure for a particular month if card bill for that particular month has been paid before.

TerenceTanWT committed:
Fixes bug usages of incorrect regex and inconsistent error messages
Resolves #209
Resolves #211

Fixes bug #209 calling incorrect regex which could cause card rebate to exceed 20%.
Fixes bug #209 inconsistent error messages.

shamussy committed:
Merge pull request #214 from TerenceTanWT/CardFix

Various major and minor bug fixes regarding Credit Card function

shamussy committed:
Add storage and commons in DG

shamussy committed:
Merge branch 'master' into loganddoc

shamussy committed:
Update build.gradle to v1.4

shamussy committed:
Fix typo in regex name

valerietyj committed:
Fix command classes, alignment issues in goals

shamussy committed:
Fix auto indentation of intellij for DG

valerietyj committed:
Merge branch 'master' into goalsfinal

valerietyj committed:
Add headers for achievements

valerietyj committed:
Edited high logic parser diagram

valerietyj committed:
Fixed checkstyle errors

shamussy committed:
Add dummy data importation class

TerenceTanWT committed:
Implemented Help Function

TerenceTanWT committed:
Fixed checkstyle issue

shamussy committed:
Fix bug in card bills for storage

BrianTKM committed:
Fix deleting card bill using /bankexpenditure

This commit prevents users from deleting card bills using /bankexpenditure and /deposit

BrianTKM committed:
Merge branch 'master' into BugFix

merge card bug fix

mong-nus committed:
Merge pull request #216 from TerenceTanWT/AddHelp

Implemented Help Function

mong-nus committed:
update PPP

mong-nus committed:
Merge branch 'master' into documentationPPP

Merge with upstream for help command

mong-nus committed:
abstract parseTypeMenu function in parseType.java

valerietyj committed:
Added Goals Sequence Diagram in Developer Guide, Updated all images for command classes, Updated DG

valerietyj committed:
Merge branch 'master' into goalsfinal

valerietyj committed:
Maintain consistency in DG Goals Design Consideration

valerietyj committed:
Fix codacy error

shamussy committed:
Merge pull request #215 from valerietyj/goalsfinal

Edited Command Classes Diagrams, Fixed misalignments for Goals

shamussy committed:
Add automatic importation of test data

valerietyj committed:
High Level Logic picture not updated

shamussy committed:
fix naming conventions

shamussy committed:
Revert newline issues

TerenceTanWT committed:
Added missing use cases and user stories for credit cards

mong-nus committed:
added logging and fix Ui issue for transaction

TerenceTanWT committed:
Merge pull request #217 from shamussy/loganddoc

Importation of test data

TerenceTanWT committed:
Added some screenshots for Card command examples

TerenceTanWT committed:
Attempt to fix incorrect image

TerenceTanWT committed:
Attempt to fix incorrect image

TerenceTanWT committed:
Fixed UG card image error

BrianTKM committed:
Fix bugs and issues raised

This commit fixes 9 digit limit
This commit fixes deleting card bill bug
This commit removes isEmpty and checks for null

BrianTKM committed:
Merge branch 'master' into BugFix

Merge with #217

BrianTKM committed:
fix github check issues

mong-nus committed:
fix Junit for everybody Ui

mong-nus committed:
improve coding style for find

TerenceTanWT committed:
Added an image

shamussy committed:
Merge pull request #218 from BrianTKM/BugFix

Bug fix

shamussy committed:
Fix dummy data to have isMature column

shamussy committed:
Update UG for /update information

valerietyj committed:
Updated UG with images, DG with manual testing. Fix bugs on reminder of goals.

valerietyj committed:
Merge branch 'master' into goalsfinal

valerietyj committed:
Remove Ui image on PPP

TerenceTanWT committed:
Fixed DeleteCardBill Storage bug

TerenceTanWT committed:
Merge branch 'master' into DGCard

Merged with master to check for conflicts

TerenceTanWT committed:
Merge branch 'FixCardBillStorage' into DGCard

Merged with FixCardBillStorage to check for conflicts

TerenceTanWT committed:
Added screenshots of example commands for CardExpenditure and CardBill

TerenceTanWT committed:
Fixed JUnit

shamussy committed:
Merge pull request #220 from TerenceTanWT/DGCard

Updated UG, DG, JUnit, and Bug Fix for DeleteCardBill Storage

shamussy committed:
Merge pull request #219 from valerietyj/goalsfinal

Updated UG and DG

shamussy committed:
Merge branch 'master' into loganddoc

shamussy committed:
Fix import data with updates from card bills

shamussy committed:
Fix readme for libraries used

shamussy committed:
Add appveyor icon

shamussy committed:
Edit DG

mong-nus committed:
added more logging for find and transfer

shamussy committed:
Add notes on manual testing

mong-nus committed:
merge with upstream

TerenceTanWT committed:
Resolves #158

BrianTKM committed:
Merge pull request #221 from shamussy/loganddoc

Fix dummy data based on latest card bill fixes

mong-nus committed:
fix Junit test

mong-nus committed:
Merge branch 'master' into documentationPPP

Merge with logging and abstraction for parseType.java with upstream

valerietyj committed:
Minor edits to goals

valerietyj committed:
Merge branch 'master' into junit

mong-nus committed:
added ppt for UG image find, transfer and card

mong-nus committed:
fix appveyor issue

mong-nus committed:
fix travis issue

shamussy committed:
Add logging for investment and bonds

mong-nus committed:
fix more appveyor issue

TerenceTanWT committed:
Merge pull request #222 from mong-nus/documentationPPP

Bug fix, upload UG picture and logging

shamussy committed:
Update logging message

shamussy committed:
Merge branch 'master' into loganddoc

TerenceTanWT committed:
Updated PPP

TerenceTanWT committed:
Merge branch 'DGCard' into PPP

BrianTKM committed:
Added pictures into the UG

shamussy committed:
Add logging for investment

BrianTKM committed:
changed JUnit to fit changes

BrianTKM committed:
added logging for transaction list

shamussy committed:
Add javadocs

TerenceTanWT committed:
Merge pull request #223 from shamussy/loganddoc

Add logging for investment and bonds

TerenceTanWT committed:
Added logging for cards

TerenceTanWT committed:
Merge branch 'master' into CardLogging

Merged master to check for conflicts

mong-nus committed:
Merge pull request #224 from BrianTKM/UGPictures

Added pictures into the UG

TerenceTanWT committed:
Implemented storage for Cards and also added JUnit for ParseRawData

TerenceTanWT committed:
Fixed checkstyle indentation issue

TerenceTanWT committed:
Fixed JUnit error

shamussy committed:
Merge pull request #225 from TerenceTanWT/CardLogging

Implemented Logging for Cards and Storage, and added JUnit for ParseRawData

BrianTKM committed:
Merge branch 'master' into Logging

Added logging

This commit adds logging for savings, expenditures, deposits and recurring expenditures

BrianTKM committed:
added logging and merged with latest

BrianTKM committed:
fix codacy issue

shamussy committed:
Merge pull request #226 from BrianTKM/Logging

Logging

BrianTKM committed:
removed tags from UG to re-tag

BrianTKM committed:
Merge branch 'master' into PPP

Merged with logging

shamussy committed:
Add loging for investment and bond parsers

valerietyj committed:
Added JUnit tests for edit profile, goals and achievements

valerietyj committed:
Merge branch 'master' into junit

valerietyj committed:
Fixed image on high level logic parser

shamussy committed:
Add Junit for parser

TerenceTanWT committed:
Updated PPP

valerietyj committed:
Added logging for goals and edit profile

valerietyj committed:
Added logging and fixed bugs and missing junit

valerietyj committed:
Merge pull request #227 from shamussy/loganddoc

Add loging for investment and bond parsers

valerietyj committed:
Merge branch 'master' into junit

valerietyj committed:
Remove unused parameter

valerietyj committed:
Fix errors in profile test cases

valerietyj committed:
Prints error message in junit

valerietyj committed:
Change regex, remove failing test case, fixed short forms

valerietyj committed:
Fix codacy errors

mong-nus committed:
update PPP UG and DG

mong-nus committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2113T-W17-3/main

Merge with upstream

valerietyj committed:
Remove declaration of logging in parse goals

mong-nus committed:
update find sequence diagram

mong-nus committed:
remove wrong pic

mong-nus committed:
upload right pic

mong-nus committed:
update numbering of header for PPP

TerenceTanWT committed:
Updated PPP

TerenceTanWT committed:
Merge branch 'master' into PPP

Check for conflicts

mong-nus committed:
update ppp and aboutus

mong-nus committed:
edit sequence diagram for find

mong-nus committed:
update PPP community contribution

TerenceTanWT committed:
Updated PPP

BrianTKM committed:
PPP

valerietyj committed:
Minor edits to developer guide and goals output

shamussy committed:
Merge pull request #228 from valerietyj/junit

Implemented Logging for Goals and Edit Profile

mong-nus committed:
added check for same bank account name for transfer

mong-nus committed:
update DG use case

TerenceTanWT committed:
Fixed various date parsing issues for card and cardbill

Fixed date parsing for card and cardbill where the program crashes when detected d/M/yyyy.
Edited printing and error messages to display MM/yyyy for card bill date instead of yyyy/MM.

BrianTKM committed:
Fix regex and update message

mong-nus committed:
amend use case in DG

mong-nus committed:
Merge pull request #229 from BrianTKM/RegexFix

Fix regex and update message

mong-nus committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2113T-W17-3/main

mong-nus committed:
Merge branch 'master' into documentationPPP

Merge with upstream

mong-nus committed:
Merge pull request #230 from TerenceTanWT/BugFixes

Fixed various date parsing issues for card and cardbill

mong-nus committed:
Merge branch 'master' of https://github.com/AY1920S1-CS2113T-W17-3/main

mong-nus committed:
Merge branch 'master' into documentationPPP

Merge with upstream for bug fix

mong-nus committed:
fix appveyor issue

TerenceTanWT committed:
Merge pull request #231 from mong-nus/documentationPPP

Bug fix and update DG and PPP

shamussy committed:
Update dg, product website

shamussy committed:
Fix codacy issues

BrianTKM committed:
Merge pull request #232 from shamussy/loganddoc

Update DG, product website

BrianTKM committed:
updated PPP

BrianTKM committed:
Merge branch 'master' into PPP

Merge with updated code

BrianTKM committed:
Corrected name errors

BrianTKM committed:
Fix PP formatting

BrianTKM committed:
removed additional line from DG

valerietyj committed:
Maintain consistency in size of sequence diagrams in DG, update format errors in UG and added missing javadocs in regex

BrianTKM committed:
fix travis issue

valerietyj committed:
Fix checkstyle errors

valerietyj committed:
Merge branch 'master' into junit

mong-nus committed:
update PPP and DG

shamussy committed:
Fix grammar errors

shamussy committed:
Merge pull request #233 from BrianTKM/PPP

PPP

shamussy committed:
Merge pull request #234 from valerietyj/junit

Minor edits to DG, UG and added missing Javadocs

mong-nus committed:
update PPP header

shamussy committed:
Fix grammar errors

mong-nus committed:
Merge branch 'master' into documentationPPP

Merge with upstream

mong-nus committed:
minor fix to PPP and DG

valerietyj committed:
Show reminder during first time run

mong-nus committed:
Merge pull request #235 from shamussy/documentationfix

Fix grammar errors

mong-nus committed:
Merge branch 'master' into documentationPPP

Merge minor fix of PPP with upstream

BrianTKM committed:
fix error message

BrianTKM committed:
Fix error message

TerenceTanWT committed:
Updated PPP, DG, and fixed printing inconsistencies for card bill

mong-nus committed:
minor changes to PPP

shamussy committed:
Fix PPP Grammar

mong-nus committed:
edit PPP

mong-nus committed:
edit PPP community documentation

mong-nus committed:
added logger for new transfer function

TerenceTanWT committed:
Fixed merge conflicts

BrianTKM committed:
removed trailing white space

BrianTKM committed:
Merge pull request #238 from shamussy/documentationfix

Fix PPP Grammar

valerietyj committed:
Fix bugs in reminders, show reminders during first time log in and fix error message

valerietyj committed:
Fix checkstyle errors

mong-nus committed:
change to PPP indentation

BrianTKM committed:
fix add bond not case-insensitive

valerietyj committed:
Merge pull request #236 from mong-nus/documentationPPP

Update DG and PPP

valerietyj committed:
Merge branch 'master' into final

BrianTKM committed:
fix DG errors

shamussy committed:
Fix manual testing bug

shamussy committed:
Merge pull request #239 from TerenceTanWT/PPP

Updated DG, PPP, and fixed UI inconsistencies in card bills

BrianTKM committed:
removed extra words in DG

valerietyj committed:
Merge branch 'master' into final

TerenceTanWT committed:
Fixed UG error

TerenceTanWT committed:
Merge branch 'master' into PPP

Merge master to check for conflicts

shamussy committed:
Merge pull request #237 from BrianTKM/bugs

Fix error message

BrianTKM committed:
Merge pull request #241 from shamussy/documentationfix

Fix manual testing bug

shamussy committed:
Merge pull request #242 from TerenceTanWT/PPP

Fixed error in UG regrading delete cardbill syntax

TerenceTanWT committed:
Updated DG manual testing

BrianTKM committed:
fix find card paid expenditure bug

BrianTKM committed:
Merge branch 'master' into bugs

Merge with latest

shamussy committed:
Merge pull request #243 from TerenceTanWT/PPP

Updated DG manual testing to update cardbill date

valerietyj committed:
Fix bugs when edit amount, error message and add missing javadocs

valerietyj committed:
Amended error birthday

shamussy committed:
Merge pull request #244 from BrianTKM/bugs

Bugs

valerietyj committed:
Merge branch 'master' into final

valerietyj committed:
fix + to be on new line

BrianTKM committed:
Merge pull request #240 from valerietyj/final

Fix bugs and error message

shamussy committed:
Disable debugging messages for CI

shamussy committed:
Fix UG markup bug

BrianTKM committed:
Merge pull request #245 from shamussy/cleanup

Final Cleanup
